proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1108 @@
+Haskell, but there's little telling what kind of havoc it would wreak
+on multiplicity inference in its current state. So we'd rather keep
+this extension for a later time.
+
+Dupable classes
+^^^^^^^^^^^^^^^
+
+We specified dupable type classes in a way that make all dupable type
+classes isomorphic. The reason for the design is that what we actually
+have in mind in the long run is to be able to add unrestricted methods
+to the type class as in the alternative above. This design has the
+advantage that it's reasonably easy to implement, only wiring in a
+type (``LinearityToken```) and two functions
+(``consumeLinearityToken`` and ``dup2LinearityToken``). Nevertheless
+there are a number of other ways to go about dupable classes.
+
+Leave it out
+~~~~~~~~~~~~
+
+Let us point out that while the ability to form a ``Linearly``
+constraint is both quite useful and absolutely at home in this
+proposal, it's also perfectly consistent to make a linear constraints
+proposal without any dupable type classes. If the design of this
+feature proves too controversial, it's definitely an option to simply
+excise the dupable class feature from the proposal.
+
+Kind of LinearityToken
+~~~~~~~~~~~~~~~~~~~~~~
+
+The proposal deliberately specifies that ``LinearityToken`` be 0
+width, and the ``dup2`` and ``consume`` functions to correspondingly
+return unboxed tuples. This is meant to emphasise that this is all for
+low level manipulation and making sure that there is no cost in
+storing ``LinearityToken``. These are all meant to be used while
+defining a dupable type class and its API, but it's not intended for
+them to be apparent in said API.
+
+An somewhat middle-ground option is to expose
+
+::
+
+  data LinearityToken# :: ZeroBitType
+
+  data LinearityToken = MkLinearityToken LinearlyToken#
+
+  -- Both functions below can be defined as easily inlineable thin
+  -- wrapper so that in most cases no allocation is needed
+  consumeLinearityToken :: LinearityToken -> ()
+  dup2LinearityToken :: LinearityToken -> (LinearityToken, LinearityToken)
+
+Just linearly
+~~~~~~~~~~~~~
+
+The most useful dupable type class in the proximate future (in fact
+the only known example yet), is ``Linearly``. So another option, to
+avoid introducing any ad hoc syntax is to simply expose the (abstract)
+``Linearly`` constraint from the ``GHC.Constraint.Linear`` module *and
+nothing else*. So that ``GHC.Constraint.Linear`` would be
+
+::
+
+  module GHC.Constraint.Linear where
+
+  -- Magically dupable
+  class Linearly
+
+  linearly'sToken :: Linearly %1 => LinearityToken
+
+  data LinearityToken :: ZeroBitType
+
+  consumeLinearityToken :: LinearityToken -> (# #)
+  dup2LinearityToken :: LinearityToken -> (# LinearityToken, LinearityToken #)
+
+
+This is quite economical from a language extension perspective, but
+the authors of this proposal are somewhat worried of the difficulties
+of wiring in a type class.
```

Comment:
To me, this alternative seems like the best design, especially given its forward-compability. However, given the potential difficulties in wiring in a type-class, I'm willing to go along with the current design, too.